Define CNB_BUILD_CONFIG_DIR behavior (#345)
buildpacks/spec: buildpack.md

Executable: `/bin/detect`, Working Dir: `<app[AR]>`

| Input                    | Attributes | Description                                       |
|--------------------------|------------|---------------------------------------------------|
| `$0`                     |            | Absolute path of `/bin/detect` executable         |
| `$CNB_BUILD_PLAN_PATH`   | E          | Absolute path of the build plan                   |
| `$CNB_BUILDPACK_DIR`     | ER         | Absolute path of the buildpack root directory     |
| `$CNB_PLATFORM_DIR`      | AR         | Absolute path of the platform directory           |
| `$CNB_PLATFORM_DIR/env/` | AR         | User-provided environment variables for build     |
| `$CNB_PLATFORM_DIR/#`    | AR         | Platform-specific extensions                      |

> | `$CNB_BUILD_CONFIG_DIR`  | AR         | Operator-provided environment variables for build |

Author: Removed

The selected group MUST be filtered to only include extensions and buildpacks with exit status zero.
The order of the extensions and buildpacks in the group MUST otherwise be preserved.

> For each `/bin/detect` executable in each buildpack, the lifecycle:

Reviewer: ```suggestion
For each `/bin/detect` executable in each extension or buildpack, the lifecycle:
```

Author: Done

> For each `/bin/detect` executable in each buildpack, the lifecycle:

Reviewer: ```suggestion
For each `/bin/detect` executable in each extension or buildpack, the lifecycle:
```

Reviewer: But TBH I don't know if this statement belongs here - we don't have it for user-provided env vars, for example.

| [†](README.md#linux-only)`LD_LIBRARY_PATH` | `/lib`       | shared libraries | [x]   | [x]    |
| [†](README.md#linux-only)`LIBRARY_PATH`    | `/lib`       | static libraries | [x]   |        |
| `CPATH`                                    | `/include`   | header files     | [x]   |        |

> | `PKG_BUILD_CONFIG_PATH`                          | `/pkgconfig` | pc files         | [x]   |        |

Reviewer 2: Unintended change?

Author: D'oh.  Fixed now.

| `CPATH`                                    | `/include`   | header files     | [x]   |        |
| `PKG_CONFIG_PATH`                          | `/pkgconfig` | pc files         | [x]   |        |

### Provided by the Platform

The following additional environment variables MUST NOT be overridden by the lifecycle.

| Env Variable           | Description                                       | Detect | Build | Launch |
|------------------------|---------------------------------------------------|--------|-------|--------|
| `CNB_STACK_ID`         | Chosen stack ID                                   | [x]    | [x]   |        |
| `BP_*`                 | User-provided variable for buildpack              | [x]    | [x]   |        |
| `BPL_*`                | User-provided variable for exec.d                 |        |       | [x]    |
| `HOME`                 | Current user's home directory                     | [x]    | [x]   | [x]    |

> | `CNB_BUILD_CONFIG_DIR` | Operator-provided environment variables for build | [x]    | [x]   |        |

Reviewer 2: We should remove this last bit, this is not part of the contract.

During the detection and build phases, the lifecycle MUST provide any user-provided environment variables as files in `<platform>/env/` with file names and contents matching the environment variable names and contents.

> During the detection and build phases, if `$CNB_BUILD_CONFIG_DIR` contains a directory path, then environment variables sourced from this directory are applied _after_ processing user-provided platform environment variables.  Environment varaibles defined in `$CNB_BUIILD_CONFIG_DIR` are operator-provided environment variables.

Reviewer: ```suggestion
During the detection and build phases, the lifecycle MUST provide as environment variables any user-provided files in `<platform>/env/` with environment variable names and contents matching the file names and contents.

During the detection and build phases, the lifecycle MUST provide as environment variables any operator-provided files in `<build-config>/env` with environment variable names and contents matching the file names and contents. This applies for all values of `clear-env` or if `clear-env` is undefined in `buildpack.toml`.

```

Reviewer: (and delete the paragraph on line 802, which doesn't make much sense to me TBH)

Author: Done

> For all values of `clear-env` or if `clear-env` is undefined in `buildpack.toml`, lifecylce MUST set operator-provided environment variables in the environment of `/bin/detect` or `/bin/build`.

Reviewer: ```suggestion
```

(if we like the suggestion in https://github.com/buildpacks/spec/pull/345/files#r1061830670)

Author: Done
